Build failure: osquery-5.12.2

### Steps To Reproduce

Steps to reproduce the behavior:
`nix build nixpkgs/50d494e5a4370394af0e09db2a2e864f146ff3b6#legacyPackages.x86_64-linux.osquery`

### Build log

```
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: /nix/store/pglcybidbkf4pc3fkiqmjspsfr2xs7h5-osquery-toolchain-bin/usr/bin/clang
-- Check for working C compiler: /nix/store/pglcybidbkf4pc3fkiqmjspsfr2xs7h5-osquery-toolchain-bin/usr/bin/clang - broken
CMake Error at /nix/store/aqckch626lg0vxh41dabyzrq0jx7gdk5-cmake-3.29.6/share/cmake-3.29/Modules/CMakeTestCCompiler.cmake:67 (message):
  The C compiler

    "/nix/store/pglcybidbkf4pc3fkiqmjspsfr2xs7h5-osquery-toolchain-bin/usr/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/build/source/build/CMakeFiles/CMakeScratch/TryCompile-Z4uLEz'
    
    Run Build Command(s): /nix/store/4mwzwdhdiqviz9bsb4i8ngy5j04qkvf5-ninja-1.12.1/bin/ninja -v cmTC_8420a
    [1/2] /nix/store/pglcybidbkf4pc3fkiqmjspsfr2xs7h5-osquery-toolchain-bin/usr/bin/clang    -o CMakeFiles/cmTC_8420a.dir/testCCompiler.c.o -c /build/source/build/CMakeFiles/CMakeScratch/TryCompile-Z4uLEz/testCCompiler.c
    FAILED: CMakeFiles/cmTC_8420a.dir/testCCompiler.c.o 
    /nix/store/pglcybidbkf4pc3fkiqmjspsfr2xs7h5-osquery-toolchain-bin/usr/bin/clang    -o CMakeFiles/cmTC_8420a.dir/testCCompiler.c.o -c /build/source/build/CMakeFiles/CMakeScratch/TryCompile-Z4uLEz/testCCompiler.c
    Inconsistency detected by ld.so: dl-call-libc-early-init.c: 37: _dl_call_libc_early_init: Assertion `sym != NULL' failed!
    ninja: build stopped: subcommand failed.
```
Full log: https://gist.github.com/sgrb/671b2b571d6a38c2755372ce38116906

### Additional context

Add any other context about the problem here.

### Notify maintainers

@znewman01 @nlewo @squalus
### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, NixOS, 24.11 (Vicuna), 24.11.20240708.655a58a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.4`
 - channels(root): `""`
 - channels(slava): `""`
 - nixpkgs: `/etc/nix/nixpkgs`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/build-support/setup-hooks/auto-patchelf.py
+++ b/pkgs/build-support/setup-hooks/auto-patchelf.py
@@ -197,7 +197,7 @@ class Dependency:
     found: bool = False     # Whether it was found somewhere
 
 
-def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list[Path] = [], extra_args: list[str] = []) -> list[Dependency]:
+def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list[Path] = [], keep_libc: bool = False, extra_args: list[str] = []) -> list[Dependency]:
     try:
         with open_elf(path) as elf:
 
@@ -267,7 +267,12 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list
             # add the dependency to rpath, as the original binary
             # presumably had it and this should be preserved.
 
+            is_libc = (libc_lib / candidate).is_file()
+
             if candidate.is_absolute() and candidate.is_file():
+                was_found = True
+                break
+            elif is_libc and not keep_libc:
                 was_found = True
                 break
             elif found_dependency := find_dependency(candidate.name, file_arch, file_osabi):
@@ -276,7 +281,7 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list
                 print(f"    {candidate} -> found: {found_dependency}")
                 was_found = True
                 break
-            elif (libc_lib / candidate).is_file():
+            elif is_libc and keep_libc:
                 was_found = True
                 break
 
@@ -306,6 +311,7 @@ def auto_patchelf(
         recursive: bool = True,
         ignore_missing: list[str] = [],
         append_rpaths: list[Path] = [],
+        keep_libc: bool = False,
         extra_args: list[str] = []) -> None:
 
     if not paths_to_patch:
@@ -319,7 +325,7 @@ def auto_patchelf(
     dependencies = []
     for path in chain.from_iterable(glob(p, '*', recursive) for p in paths_to_patch):
         if not path.is_symlink() and path.is_file():
-            dependencies += auto_patchelf_file(path, runtime_deps, append_rpaths, extra_args)
+            dependencies += auto_patchelf_file(path, runtime_deps, append_rpaths, keep_libc, extra_args)
 
     missing = [dep for dep in dependencies if not dep.found]
 
@@ -378,6 +384,12 @@ def main() -> None:
         help="Paths to append to all runtime paths unconditionally",
     )
     parser.add_argument(
+        "--keep-libc",
+        dest="keep_libc",
+        action="store_true",
+        help="Leave libc in the rpath if it's found",
+    )
+    parser.add_argument(
         "--extra-args",
         # Undocumented Python argparse feature: consume all remaining arguments
         # as values for this one. This means this argument should always be passed
@@ -398,6 +410,7 @@ def main() -> None:
         args.recursive,
         args.ignore_missing,
         append_rpaths=args.append_rpaths,
+        keep_libc=args.keep_libc,
         extra_args=args.extra_args)
 
 

--- a/pkgs/build-support/setup-hooks/auto-patchelf.py
+++ b/pkgs/build-support/setup-hooks/auto-patchelf.py
@@ -257,20 +257,20 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list
             # 1. If a candidate is an absolute path, it is already a
             #    valid dependency if that path exists, and nothing needs
             #    to be done. It should be an error if that path does not exist.
-            # 2. If a candidate is found in our library dependencies, that
-            #    dependency should be added to rpath. This search is skipped
-            #    for libc dependencies when keep_libc is False.
-            # 3. If a candidate is found in libc, it will be correctly
-            #    resolved by the dynamic linker automatically.
+            # 2. If a candidate is found within libc, it should be dropped
+            #    and resolved automatically by the dynamic linker, unless
+            #    keep_libc is enabled.
+            # 3. If a candidate is found in our library dependencies, that
+            #    dependency should be added to rpath.
+            # 4. If all of the above fail, libc dependencies should still be
+            #    considered found. This is in contrast to step 2, because
+            #    enabling keep_libc should allow libc to be found in step 3
+            #    if possible to preserve its presence in rpath.
             #
             # These conditions are checked in this order, because #2
             # and #3 may both be true. In that case, we still want to
             # add the dependency to rpath, as the original binary
             # presumably had it and this should be preserved.
-            #
-            # The keep_libc flag affects how libc dependencies are handled. If
-            # it's True, the loop will attempt to search for and relink them.
-            # When it's False, they are removed from the rpath.
 
             is_libc = (libc_lib / candidate).is_file()
 

--- a/pkgs/build-support/setup-hooks/auto-patchelf.py
+++ b/pkgs/build-support/setup-hooks/auto-patchelf.py
@@ -258,7 +258,8 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list
             #    valid dependency if that path exists, and nothing needs
             #    to be done. It should be an error if that path does not exist.
             # 2. If a candidate is found in our library dependencies, that
-            #    dependency should be added to rpath.
+            #    dependency should be added to rpath. This search is skipped
+            #    for libc dependencies when keep_libc is False.
             # 3. If a candidate is found in libc, it will be correctly
             #    resolved by the dynamic linker automatically.
             #
@@ -266,6 +267,10 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: list
             # and #3 may both be true. In that case, we still want to
             # add the dependency to rpath, as the original binary
             # presumably had it and this should be preserved.
+            #
+            # The keep_libc flag affects how libc dependencies are handled. If
+            # it's True, the loop will attempt to search for and relink them.
+            # When it's False, they are removed from the rpath.
 
             is_libc = (libc_lib / candidate).is_file()
 
@@ -387,7 +392,7 @@ def main() -> None:
         "--keep-libc",
         dest="keep_libc",
         action="store_true",
-        help="Leave libc in the rpath if it's found",
+        help="Attempt to search for and relink libc dependencies.",
     )
     parser.add_argument(
         "--extra-args",

--- a/pkgs/build-support/setup-hooks/auto-patchelf.sh
+++ b/pkgs/build-support/setup-hooks/auto-patchelf.sh
@@ -58,11 +58,13 @@ autoPatchelf() {
         local appendRunpathsArray=( "${appendRunpaths[@]}" )
         local runtimeDependenciesArray=( "${runtimeDependencies[@]}" )
         local patchelfFlagsArray=( "${patchelfFlags[@]}" )
+        local autoPatchelfFlagsArray=( "${autoPatchelfFlags[@]}" )
     else
         readarray -td' ' ignoreMissingDepsArray < <(echo -n "$autoPatchelfIgnoreMissingDeps")
         local appendRunpathsArray=($appendRunpaths)
         local runtimeDependenciesArray=($runtimeDependencies)
         local patchelfFlagsArray=($patchelfFlags)
+        local autoPatchelfFlagsArray=($autoPatchelfFlags)
     fi
 
     # Check if ignoreMissingDepsArray contains "1" and if so, replace it with
@@ -85,6 +87,7 @@ autoPatchelf() {
                "${extraAutoPatchelfLibs[@]}"                            \
         --runtime-dependencies "${runtimeDependenciesArray[@]/%//lib}"  \
         --append-rpaths "${appendRunpathsArray[@]}"                     \
+        "${autoPatchelfFlagsArray[@]}"                                  \
         --extra-args "${patchelfFlagsArray[@]}"
 }
 

--- a/pkgs/os-specific/linux/systemd/default.nix
+++ b/pkgs/os-specific/linux/systemd/default.nix
@@ -342,6 +342,8 @@ stdenv.mkDerivation (finalAttrs: {
     ]
   ;
 
+  autoPatchelfFlags = [ "--keep-libc" ];
+
   buildInputs =
     [
       libxcrypt

